File: schema_salad/typescript/util/loaders/RecordLoader.ts
import { Saveable, Loader, LoadingOptions, TypeGuards, ValidationException } from '../Internal'

export class _RecordLoader implements Loader {
  creatorFunc: (doc: any, baseuri: string, loadingOptions: LoadingOptions, docRoot?: string) => Promise<Saveable>
  container?: string

  constructor (createrFunc: (doc: any, baseuri: string, loadingOptions: LoadingOptions, docRoot?: string) => Promise<Saveable>, container?: string) {
    this.creatorFunc = createrFunc
    this.container = container
  }

  async load (doc: any, baseuri: string, loadingOptions: LoadingOptions, docRoot?: string): Promise<Saveable> {
    if (!TypeGuards.isDictionary(doc)) {
      throw new ValidationException('Expected a dict')
    }
    if (this.container !== undefined) {
      loadingOptions = new LoadingOptions({ copyFrom: loadingOptions, container: this.container })
    }
    return await this.creatorFunc(doc, baseuri, loadingOptions, docRoot)
  }
}

File: schema_salad/dotnet/util/Loaders/RecordLoader.cs
﻿using System.Collections;

namespace ${project_name};

internal class RecordLoader<T> : ILoader<T> where T : ISaveable
{
    private readonly string? container;

    public RecordLoader(in string? container)
    {
        this.container = container;
    }

    public T Load(in object doc, in string baseUri, in LoadingOptions loadingOptions, in string? docRoot = null)
    {
        if (doc is not IDictionary)
        {
            throw new ValidationException($"Expected object with type of Dictionary but got {doc.GetType()}");
        }

        LoadingOptions innerLoadingOptions = loadingOptions;
        if (this.container != null)
        {
            innerLoadingOptions = new LoadingOptions(copyFrom: loadingOptions, container: this.container);
        }

        return (T)T.FromDoc(doc, baseUri, innerLoadingOptions, docRoot);
    }

    object ILoader.Load(in object doc, in string baseuri, in LoadingOptions loadingOptions, in string? docRoot)
    {
        return Load(doc,
                    baseuri,
                    loadingOptions,
                    docRoot)!;
    }
}

